File: android/app/src/main/kotlin/app/chronicler/Downloads.kt
package app.chronicler

import android.content.Context
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File

// A book whose chapters have been downloaded: enough metadata to list and play it with
// the server unreachable. Written to manifest.json beside the audio files.
@Serializable
data class DownloadedBook(val book: Book, val chapters: List<Chapter>)

// Chapter progress kept on disk so offline playback resumes in the right place.
// `dirty` marks progress recorded while the server was unreachable; it gets pushed on
// the next load that reaches the API.
@Serializable
data class StoredProgress(
    val positionSeconds: Double = 0.0,
    val durationSeconds: Double = 0.0,
    val isListened: Boolean = false,
    val dirty: Boolean = false,
)

// Offline downloads: stores chapter audio under filesDir/downloads/<chapterId>.audio, the
// book + chapter metadata in manifest.json, the cover in cover-<bookId>.img and chapter
// progress in progress.json — so the whole offline library works with no network at all.
object Downloads {
    private val http = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val lock = Any()

    private fun dir(context: Context): File =
        File(context.filesDir, "downloads").apply { mkdirs() }

    fun file(context: Context, chapterId: Int): File = File(dir(context), "$chapterId.audio")

    fun isDownloaded(context: Context, chapterId: Int): Boolean =
        file(context, chapterId).let { it.exists() && it.length() > 0 }

    /** Local file URI if downloaded, else the streaming URL. */
    fun sourceUri(context: Context, chapterId: Int, streamUrl: String): String =
        if (isDownloaded(context, chapterId)) Uri.fromFile(file(context, chapterId)).toString() else streamUrl

    suspend fun download(context: Context, chapterId: Int, url: String): Boolean =
        withContext(Dispatchers.IO) {
            runCatching {
                val resp = http.newCall(Request.Builder().url(url).build()).execute()
                resp.use {
                    if (!it.isSuccessful) return@withContext false
                    val tmp = File(dir(context), "$chapterId.part")
                    it.body?.byteStream()?.use { input ->
                        tmp.outputStream().use { out -> input.copyTo(out) }
                    } ?: return@withContext false
                    if (tmp.length() <= 0) { tmp.delete(); return@withContext false }
                    tmp.renameTo(file(context, chapterId))
                }
                true
            }.getOrDefault(false)
        }

    fun deleteChapter(context: Context, chapterId: Int) {
        file(context, chapterId).delete()
    }

    fun deleteBook(context: Context, chapterIds: List<Int>) {
        chapterIds.forEach { deleteChapter(context, it) }
    }

    suspend fun totalSize(context: Context): Long = withContext(Dispatchers.IO) {
        dir(context).listFiles()?.filter { it.name.endsWith(".audio") }
            ?.sumOf { it.length() } ?: 0L
    }

    // ── Manifest (which books/chapters the downloads belong to) ──

    private fun manifestFile(context: Context) = File(dir(context), "manifest.json")

    private fun readManifest(context: Context): MutableMap<String, DownloadedBook> = runCatching {
        val f = manifestFile(context)
        if (!f.exists()) return@runCatching mutableMapOf<String, DownloadedBook>()
        json.decodeFromString<Map<String, DownloadedBook>>(f.readText()).toMutableMap()
    }.getOrDefault(mutableMapOf())

    private fun writeManifest(context: Context, manifest: Map<String, DownloadedBook>) {
        runCatching { manifestFile(context).writeText(json.encodeToString<Map<String, DownloadedBook>>(manifest)) }
    }

    /** Remembers a book + its chapter list so the offline library can list and play it. */
    fun record(context: Context, book: Book, chapters: List<Chapter>) {
        if (chapters.isEmpty()) return
        synchronized(lock) {
            val manifest = readManifest(context)
            manifest[book.id.toString()] = DownloadedBook(book, chapters)
            writeManifest(context, manifest)
        }
    }

    fun entry(context: Context, bookId: Int): DownloadedBook? =
        synchronized(lock) { readManifest(context)[bookId.toString()] }

    /** Books with at least one chapter still on disk, A→Z by title. */
    fun downloadedBooks(context: Context): List<DownloadedBook> =
        synchronized(lock) { readManifest(context).values.toList() }
            .filter { entry -> entry.chapters.any { isDownloaded(context, it.id) } }
            .sortedBy { it.book.title.lowercase() }

    /** Cheap check: is anything downloaded at all? */
    fun hasAny(context: Context): Boolean =
        dir(context).listFiles()?.any { it.name.endsWith(".audio") && it.length() > 0 } ?: false

    fun downloadedCount(context: Context, bookId: Int): Int =
        entry(context, bookId)?.chapters?.count { isDownloaded(context, it.id) } ?: 0

    fun bookSize(context: Context, bookId: Int): Long =
        entry(context, bookId)?.chapters?.sumOf { file(context, it.id).length() } ?: 0L

    /** Deletes every downloaded chapter of a book plus its cached cover + manifest entry. */
    fun purgeBook(context: Context, bookId: Int) {
        entry(context, bookId)?.chapters?.forEach { deleteChapter(context, it.id) }
        coverFile(context, bookId).delete()
        synchronized(lock) {
            val manifest = readManifest(context)
            manifest.remove(bookId.toString())
            writeManifest(context, manifest)
        }
    }

    /** Drops the manifest entry once the last chapter of a book has been removed. */
    fun pruneIfEmpty(context: Context, bookId: Int) {
        val entry = entry(context, bookId) ?: return
        if (entry.chapters.none { isDownloaded(context, it.id) }) purgeBook(context, bookId)
    }

    // ── Cover (kept beside the audio so the offline library isn't a wall of placeholders) ──

    fun coverFile(context: Context, bookId: Int): File = File(dir(context), "cover-$bookId.img")

    fun coverBytes(context: Context, bookId: Int): ByteArray? =
        coverFile(context, bookId).let { if (it.exists() && it.length() > 0) it.readBytes() else null }

    /** Fetches and stores the cover for a downloaded book (no-op if it can't be fetched). */
    suspend fun cacheCover(context: Context, bookId: Int, api: ApiClient) {
        val bytes = api.coverBytes(bookId) ?: return
        withContext(Dispatchers.IO) { runCatching { coverFile(context, bookId).writeBytes(bytes) } }
    }

    // ── Local chapter progress ──

    private fun progressFile(context: Context) = File(dir(context), "progress.json")

    private fun readProgress(context: Context): MutableMap<String, StoredProgress> = runCatching {
        val f = progressFile(context)
        if (!f.exists()) return@runCatching mutableMapOf<String, StoredProgress>()
        json.decodeFromString<Map<String, StoredProgress>>(f.readText()).toMutableMap()
    }.getOrDefault(mutableMapOf())

    private fun writeProgress(context: Context, progress: Map<String, StoredProgress>) {
        runCatching { progressFile(context).writeText(json.encodeToString<Map<String, StoredProgress>>(progress)) }
    }

    fun localProgress(context: Context, chapterId: Int): StoredProgress? =
        synchronized(lock) { readProgress(context)[chapterId.toString()] }

    fun setLocalProgress(context: Context, chapterId: Int, position: Double, duration: Double,
                         isListened: Boolean, dirty: Boolean) {
        synchronized(lock) {
            val progress = readProgress(context)
            val existing = progress[chapterId.toString()]
            progress[chapterId.toString()] = StoredProgress(
                positionSeconds = position,
                durationSeconds = if (duration > 0) duration else (existing?.durationSeconds ?: 0.0),
                isListened = isListened,
                dirty = dirty,
            )
            writeProgress(context, progress)
        }
    }

    /** Progress recorded while offline, keyed by chapter id. */
    fun dirtyProgress(context: Context): Map<Int, StoredProgress> =
        synchronized(lock) { readProgress(context) }
            .filterValues { it.dirty }
            .mapNotNull { (k, v) -> k.toIntOrNull()?.let { it to v } }
            .toMap()

    fun clearDirty(context: Context, chapterId: Int) {
        synchronized(lock) {
            val progress = readProgress(context)
            progress[chapterId.toString()]?.let {
                progress[chapterId.toString()] = it.copy(dirty = false)
                writeProgress(context, progress)
            }
        }
    }
}

File: Chronicler/Chronicler/Downloads.swift
import Foundation

// A book whose chapters have been downloaded: enough metadata to list and play it
// with the server unreachable. Written to manifest.json beside the audio files.
struct DownloadedBook: Codable, Identifiable, Hashable {
    var book: Book
    var chapters: [Chapter]
    var id: Int { book.id }
}

// Chapter progress kept on disk so offline playback resumes in the right place.
// `dirty` marks progress recorded while the server was unreachable; it gets pushed
// on the next load that reaches the API.
struct StoredProgress: Codable {
    var positionSeconds: Double = 0
    var durationSeconds: Double = 0
    var isListened: Bool = false
    var dirty: Bool = false
}

// Offline downloads: stores chapter audio under downloads/<chapterId>.audio, the book +
// chapter metadata in manifest.json, the cover in cover-<bookId>.img, and chapter progress
// in progress.json — so the whole offline library works without a single network call.
// Swift port of Android's Downloads object.
enum Downloads {
    private static let lock = NSLock()

    // Downloads live in Application Support: Caches can be evicted by iOS under storage
    // pressure, which is exactly when you'd want your offline books. Anything left in the
    // old Caches/downloads location is moved over on first use.
    private static let root: URL = {
        let fm = FileManager.default
        let base = (try? fm.url(for: .applicationSupportDirectory, in: .userDomainMask,
                                appropriateFor: nil, create: true))
            ?? fm.urls(for: .cachesDirectory, in: .userDomainMask)[0]
        var dir = base.appendingPathComponent("downloads", isDirectory: true)
        try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        var values = URLResourceValues()
        values.isExcludedFromBackup = true          // re-downloadable content
        try? dir.setResourceValues(values)

        let old = fm.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("downloads", isDirectory: true)
        if old.path != dir.path,
           let items = try? fm.contentsOfDirectory(at: old, includingPropertiesForKeys: nil) {
            for item in items {
                let dest = dir.appendingPathComponent(item.lastPathComponent)
                if !fm.fileExists(atPath: dest.path) { try? fm.moveItem(at: item, to: dest) }
            }
            try? fm.removeItem(at: old)
        }

        // Older builds saved every download as "<id>.audio", which AVFoundation can't
        // type and so plays silence. Rename them to the common case (mp3).
        if let items = try? fm.contentsOfDirectory(at: dir, includingPropertiesForKeys: nil) {
            for item in items where item.pathExtension == "audio" {
                let renamed = item.deletingPathExtension().appendingPathExtension("mp3")
                if !fm.fileExists(atPath: renamed.path) { try? fm.moveItem(at: item, to: renamed) }
            }
        }
        return dir
    }()

    private static func dir() -> URL { root }

    /// The audio file on disk for a chapter, whatever extension it was saved with.
    static func file(chapterId: Int) -> URL? {
        let items = (try? FileManager.default.contentsOfDirectory(at: dir(),
                                                                  includingPropertiesForKeys: nil)) ?? []
        return items.first { $0.deletingPathExtension().lastPathComponent == String(chapterId) }
    }

    // AVFoundation picks a decoder off the file extension, so a download saved as a
    // generic ".audio" silently plays nothing. Map the server's mime onto a real one.
    static func fileExtension(for mimeType: String?) -> String {
        switch (mimeType ?? "").lowercased() {
        case let m where m.contains("mp4") || m.contains("m4a") || m.contains("m4b") || m.contains("aac"):
            return "m4a"
        case let m where m.contains("ogg") || m.contains("opus"): return "ogg"
        case let m where m.contains("flac"): return "flac"
        case let m where m.contains("wav"): return "wav"
        default: return "mp3"
        }
    }

    static func isDownloaded(chapterId: Int) -> Bool {
        guard let url = file(chapterId: chapterId),
              let size = try? FileManager.default.attributesOfItem(atPath: url.path)[.size] as? Int
        else { return false }
        return size > 0
    }

    /// Local file URL if downloaded, else the streaming URL.
    static func sourceURL(chapterId: Int, streamURL: URL) -> URL {
        isDownloaded(chapterId: chapterId) ? (file(chapterId: chapterId) ?? streamURL) : streamURL
    }

    /// Downloads a chapter's audio to disk. Returns true on success.
    static func download(chapterId: Int, url: URL, token: String?, mimeType: String? = nil) async -> Bool {
        var req = URLRequest(url: url)
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        do {
            let (tmp, resp) = try await URLSession.shared.download(for: req)
            let http = resp as? HTTPURLResponse
            guard let code = http?.statusCode, (200..<300).contains(code) else { return false }
            // Prefer what the server actually served over the chapter's declared mime.
            let ext = fileExtension(for: http?.mimeType ?? mimeType)
            deleteChapter(chapterId: chapterId)
            let dest = dir().appendingPathComponent("\(chapterId).\(ext)")
            try FileManager.default.moveItem(at: tmp, to: dest)
            return isDownloaded(chapterId: chapterId)
        } catch {
            return false
        }
    }

    static func deleteChapter(chapterId: Int) {
        if let existing = file(chapterId: chapterId) {
            try? FileManager.default.removeItem(at: existing)
        }
    }

    // ── Manifest (which books/chapters the downloads belong to) ──

    private static var manifestFile: URL { dir().appendingPathComponent("manifest.json") }

    private static func readManifest() -> [Int: DownloadedBook] {
        guard let data = try? Data(contentsOf: manifestFile),
              let raw = try? JSONDecoder().decode([String: DownloadedBook].self, from: data)
        else { return [:] }
        return Dictionary(uniqueKeysWithValues: raw.compactMap { key, value in
            Int(key).map { ($0, value) }
        })
    }

    private static func writeManifest(_ manifest: [Int: DownloadedBook]) {
        let raw = Dictionary(uniqueKeysWithValues: manifest.map { (String($0.key), $0.value) })
        if let data = try? JSONEncoder().encode(raw) {
            try? data.write(to: manifestFile, options: .atomic)
        }
    }

    /// Remembers a book + its chapter list so the offline library can list and play it.
    static func record(book: Book, chapters: [Chapter]) {
        guard !chapters.isEmpty else { return }
        lock.lock(); defer { lock.unlock() }
        var manifest = readManifest()
        manifest[book.id] = DownloadedBook(book: book, chapters: chapters)
        writeManifest(manifest)
    }

    static func entry(bookId: Int) -> DownloadedBook? {
        lock.lock(); defer { lock.unlock() }
        return readManifest()[bookId]
    }

    /// Books with at least one chapter still on disk, A→Z by title.
    static func downloadedBooks() -> [DownloadedBook] {
        lock.lock()
        let all = readManifest()
        lock.unlock()
        return all.values
            .filter { entry in entry.chapters.contains { isDownloaded(chapterId: $0.id) } }
            .sorted { $0.book.title.lowercased() < $1.book.title.lowercased() }
    }

    static func hasAny() -> Bool {
        // Cheap check: any chapter file (named "<chapterId>.<ext>") at all.
        let items = (try? FileManager.default.contentsOfDirectory(at: dir(),
                                                                  includingPropertiesForKeys: nil)) ?? []
        return items.contains { Int($0.deletingPathExtension().lastPathComponent) != nil }
    }

    static func downloadedCount(bookId: Int) -> Int {
        entry(bookId: bookId)?.chapters.filter { isDownloaded(chapterId: $0.id) }.count ?? 0
    }

    static func size(bookId: Int) -> Int64 {
        guard let entry = entry(bookId: bookId) else { return 0 }
        return entry.chapters.reduce(0) { total, chapter in
            guard let path = file(chapterId: chapter.id)?.path,
                  let attrs = try? FileManager.default.attributesOfItem(atPath: path) else { return total }
            return total + ((attrs[.size] as? NSNumber)?.int64Value ?? 0)
        }
    }

    static func totalSize() -> Int64 {
        let items = (try? FileManager.default.contentsOfDirectory(
            at: dir(), includingPropertiesForKeys: [.fileSizeKey])) ?? []
        return items
            .filter { Int($0.deletingPathExtension().lastPathComponent) != nil }   // audio only
            .reduce(0) { $0 + Int64((try? $1.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0) }
    }

    /// Deletes every downloaded chapter of a book plus its cached cover + manifest entry.
    static func deleteBook(bookId: Int) {
        entry(bookId: bookId)?.chapters.forEach { deleteChapter(chapterId: $0.id) }
        deleteCover(bookId: bookId)
        lock.lock()
        var manifest = readManifest()
        manifest[bookId] = nil
        writeManifest(manifest)
        lock.unlock()
    }

    /// Drops the manifest entry once the last chapter of a book has been removed.
    static func pruneIfEmpty(bookId: Int) {
        guard let entry = entry(bookId: bookId) else { return }
        guard !entry.chapters.contains(where: { isDownloaded(chapterId: $0.id) }) else { return }
        deleteBook(bookId: bookId)
    }

    // ── Cover (kept beside the audio so the offline library isn't a wall of placeholders) ──

    static func coverFile(bookId: Int) -> URL {
        dir().appendingPathComponent("cover-\(bookId).img")
    }
    static func saveCover(bookId: Int, data: Data) {
        try? data.write(to: coverFile(bookId: bookId), options: .atomic)
    }
    static func coverData(bookId: Int) -> Data? {
        try? Data(contentsOf: coverFile(bookId: bookId))
    }
    static func deleteCover(bookId: Int) {
        try? FileManager.default.removeItem(at: coverFile(bookId: bookId))
    }
    /// Fetches and stores the cover for a downloaded book (no-op if it can't be fetched).
    static func cacheCover(bookId: Int, api: APIClient) async {
        guard let data = await api.coverData(bookId) else { return }
        saveCover(bookId: bookId, data: data)
    }

    // ── Local chapter progress ──

    private static var progressFile: URL { dir().appendingPathComponent("progress.json") }

    private static func readProgress() -> [String: StoredProgress] {
        guard let data = try? Data(contentsOf: progressFile),
              let raw = try? JSONDecoder().decode([String: StoredProgress].self, from: data)
        else { return [:] }
        return raw
    }

    private static func writeProgress(_ progress: [String: StoredProgress]) {
        if let data = try? JSONEncoder().encode(progress) {
            try? data.write(to: progressFile, options: .atomic)
        }
    }

    static func localProgress(chapterId: Int) -> StoredProgress? {
        lock.lock(); defer { lock.unlock() }
        return readProgress()[String(chapterId)]
    }

    static func setLocalProgress(chapterId: Int, position: Double, duration: Double,
                                 isListened: Bool, dirty: Bool) {
        lock.lock(); defer { lock.unlock() }
        var progress = readProgress()
        let existing = progress[String(chapterId)]
        progress[String(chapterId)] = StoredProgress(
            positionSeconds: position,
            durationSeconds: duration > 0 ? duration : (existing?.durationSeconds ?? 0),
            isListened: isListened,
            dirty: dirty)
        writeProgress(progress)
    }

    /// Progress recorded while offline, keyed by chapter id.
    static func dirtyProgress() -> [Int: StoredProgress] {
        lock.lock(); defer { lock.unlock() }
        return Dictionary(uniqueKeysWithValues: readProgress()
            .filter { $0.value.dirty }
            .compactMap { key, value in Int(key).map { ($0, value) } })
    }

    static func clearDirty(chapterId: Int) {
        lock.lock(); defer { lock.unlock() }
        var progress = readProgress()
        progress[String(chapterId)]?.dirty = false
        writeProgress(progress)
    }
}
